sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/GenerateProjection.scala

```diff
@@ -84,9 +84,15 @@ object GenerateProjection extends CodeGenerator[Seq[Expression], Projection] {
         val elementName = newTermName(s"c$i")
         val iLit = ru.Literal(Constant(i))
 
-        q"if(i == $ordinal) { if(isNullAt($i)) return null else return $elementName }"
+        q"if(i == $ordinal) { return $elementName }"
       }
-      q"override def apply(i: Int): Any = { ..$cases; $accessorFailure }"
+      q"""
+        override def apply(i: Int): Any = {
+          if (i < 0 || i >= this.length) $accessorFailure
```

Comment:
Yes, but you are adding an if check that will never actually be used when the optimizer is doing its job correctly.  Why optimize for user error?